src/bpfilter/xlate/cli.c

```diff
@@ -56,6 +60,206 @@ int _bf_cli_ruleset_flush(const struct bf_request *request,
+        if (!chain)
+            bf_err("Chain list element pointed to null data\n");
+
+        bf_assert(chain);
+
+        struct bf_cgen *cgen = bf_ctx_get_cgen(chain->hook, NULL);
+        if (!cgen)
+            bf_err("Got a null cgen for chain element\n");
+
+        // Iterate over each rule in the current chain
+        int map_idx = 0;
+        bf_list_foreach (&chain->rules, rule_node) {
+            struct bf_rule *rule = bf_list_node_get_data(rule_node);
+            if (!chain)
+                bf_err("Rule list element pointed to null data\n");
+
+            if (rule->counters) {
+                // Query the BPF map for the counter values
+                r = bf_cgen_get_counter(cgen, map_idx,
+                                        &counters[counter_index]);
+                if (r < 0)
+                    return bf_err_r(r, "Failed to get rule counter\n");
+            }
+
+            /* Note: The map has entries for every rule regardless of whether the
+             * input rule specified "COUNTERS" or not. Thus, advancing map_index
+             * is unconditional. When rule->counters == false, the corresponding
+             * counter value will remain {0,0}. */
+            counter_index++;
+            map_idx++;
+        }
+
+        // Chain's policy counter
+        r = bf_cgen_get_counter(cgen, BF_COUNTER_POLICY,
+                                &counters[counter_index]);
+        if (r < 0)
+            return bf_err_r(r, "Failed to get policy counter\n");
+        counter_index++;
+
+        // Chain's error counter
+        r = bf_cgen_get_counter(cgen, BF_COUNTER_ERRORS,
+                                &counters[counter_index]);
+        if (r < 0)
+            return bf_err_r(r, "Failed to get error counter\n");
+    }
+
+    return 0;
+}
+
+static int _bf_cli_get_chain_list(bf_list **chain_list)
+{
+    bf_list_ops ops = {// chain_list will only contain the chains, not own them.
+                       .free = (bf_list_ops_free)bf_nop_free,
+                       .marsh = (bf_list_ops_marsh)bf_chain_marsh};
+
+    int r = bf_list_new(chain_list, &ops);
+    if (r < 0)
+        return bf_err_r(r, "Failed to create the chain list");
+
+    // Loop over codegens which correspond to hooks
+    for (int i = 0; i < _BF_HOOK_MAX; ++i) {
+        struct bf_cgen *cgen;
+
+        cgen = bf_ctx_get_cgen(i, NULL);
+
+        if (!cgen || !cgen->chain)
+            continue;
+
+        r = bf_list_add_tail(*chain_list, cgen->chain);
+        if (r)
+            return bf_err_r(r, "Failed to add chain to list");
+    }
+
+    return 0;
+}
+
+static int _bf_cli_get_counters_marsh(struct bf_marsh **counter_marsh,
+                                      bf_list *chain_list, bool with_counters)
+{
+    int r;
+    size_t num_counters = 0;
+    struct bf_counter *counters = NULL;
+
+    bf_assert(bf_list_size(chain_list) > 0);
+
+    // If we don't want counters, return an empty marsh
+    if (!with_counters) {
```

Comment:
`_bf_cli_get_counters_marsh()` should provide a marsh of the counters, not "maybe a marsh of the counters". I.e. if `with_counters` is false, this function shouldn't be called.